airbytehq/airbyte
# 🪟🎉 Connector form: Reload config if updated during connection check

## What
Fixes https://github.com/airbytehq/airbyte/issues/21041 

This PR will make sure updates to the configuration of an existing source or destination during the connection check won't be lost due to a subsequent update dispatched by the UI.

## How

The flag indicating that the config got updated during the check is already returned to the webapp (`jobInfo.connectorConfigurationUpdated`). In `airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx` this response can be checked and if it's set, the actual `update` call can be prevented as it has been done implicitly by the connection check.

At this point the form should be reloaded with the new config. This is done by passing down a `reloadConfig` callback which is calling `queryClient.invalidateQueries` for the source/destination which will cause react-query to reload the config and pass down the new state.

## How to test

Unfortunately it's hard to cause a real update-on-check to happen - I tested by mocking the server like this (need to run `./gradlew airbyte-server:assemble` and restart the docker containers):
```
--- a/airbyte-server/src/main/java/io/airbyte/server/apis/SourceApiController.java
+++ b/airbyte-server/src/main/java/io/airbyte/server/apis/SourceApiController.java
@@ -57,7 +57,11 @@ public class SourceApiController implements SourceApi {
   @Secured({EDITOR})
   @Override
   public CheckConnectionRead checkConnectionToSourceForUpdate(final SourceUpdate sourceUpdate) {
-    return ApiHelper.execute(() -> schedulerHandler.checkSourceConnectionFromSourceIdForUpdate(sourceUpdate));
+    return ApiHelper.execute(() -> {
+      final CheckConnectionRead r = schedulerHandler.checkSourceConnectionFromSourceIdForUpdate(sourceUpdate);
+      r.setJobInfo(r.getJobInfo().connectorConfigurationUpdated(true));
+      return r;
+    });
   }

   @Post("/clone")
@@ -95,7 +99,11 @@ public class SourceApiController implements SourceApi {
   @Secured({READER})
   @Override
   public SourceRead getSource(final SourceIdRequestBody sourceIdRequestBody) {
-    return ApiHelper.execute(() -> sourceHandler.getSource(sourceIdRequestBody));
+    return ApiHelper.execute(() -> {
+      final SourceRead r = sourceHandler.getSource(sourceIdRequestBody);
+      r.setName(r.getName() + System.currentTimeMillis());
+      return r;
+    });
   }
```

`checkConnectionToSourceForUpdate` automatically sets the flag to true for all connection checks. The `getSource` endpoint adds the current time in ms to the name of the connection to make it obvious the form actually updated.

https://www.loom.com/share/7d73b667dc4b40378f268ad8bf987f22
